File: api/src/routes/payments.ts
import { Hono } from 'hono';
import { Env } from '../types';
import { requireAuth } from '../auth';

const router = new Hono<{ Bindings: Env }>();

function randomRef() {
  return 'PAY-' + Array.from(crypto.getRandomValues(new Uint8Array(4))).map(b => b.toString(16).padStart(2, '0')).join('').toUpperCase();
}

router.post('/', async (c) => {
  const user = await requireAuth(c.req.header('authorization'), c.env, c.env.DB);
  const { amount, method, phone_number, description } = await c.req.json();
  const ref = randomRef();
  const { meta } = await c.env.DB.prepare(
    'INSERT INTO payments (user_id, amount, currency, method, status, reference, phone_number, description) VALUES (?, ?, ?, ?, ?, ?, ?, ?)'
  ).bind(user.id, amount, 'KES', method || 'mpesa', 'completed', ref, phone_number || null, description || null).run();
  const payment = await c.env.DB.prepare('SELECT id, reference, amount, status, mpesa_code FROM payments WHERE id = ?').bind(meta.last_row_id).first();
  return c.json({ message: 'Payment completed', payment });
});

router.get('/', async (c) => {
  const user = await requireAuth(c.req.header('authorization'), c.env, c.env.DB);
  const result = await c.env.DB.prepare('SELECT id, reference, amount, currency, method, status, description, created_at FROM payments WHERE user_id = ? ORDER BY created_at DESC').bind(user.id).all();
  return c.json(result.results);
});

export default router;

File: backend/app/routes/payments.py
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.orm import Session
from app.database import get_db
from app.models.payment import Payment, PaymentStatus, PaymentMethod
from app.models.user import User
from app.utils.auth import get_current_user

router = APIRouter(prefix="/api/payments", tags=["Payments"])

class PaymentRequest(BaseModel):
    amount: float
    method: str = "mpesa"
    phone_number: str | None = None
    description: str | None = None

@router.post("/")
def create_payment(req: PaymentRequest, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    import uuid
    payment = Payment(
        user_id=current_user.id,
        amount=req.amount,
        currency="KES",
        method=PaymentMethod(req.method),
        status=PaymentStatus.PENDING,
        reference=f"PAY-{uuid.uuid4().hex[:8].upper()}",
        phone_number=req.phone_number,
        description=req.description,
    )
    db.add(payment)
    db.commit()
    db.refresh(payment)
    # In production, initiate M-Pesa STK Push here
    payment.status = PaymentStatus.COMPLETED
    payment.mpesa_code = f"MPESA{payment.reference}"
    db.commit()
    return {
        "message": "Payment initiated",
        "payment": {
            "id": payment.id,
            "reference": payment.reference,
            "amount": payment.amount,
            "status": payment.status.value,
            "mpesa_code": payment.mpesa_code,
        }
    }

@router.get("/")
def list_payments(current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    payments = db.query(Payment).filter(Payment.user_id == current_user.id).all()
    return [
        {
            "id": p.id,
            "reference": p.reference,
            "amount": p.amount,
            "currency": p.currency,
            "method": p.method.value if p.method else None,
            "status": p.status.value,
            "description": p.description,
            "created_at": str(p.created_at),
        }
        for p in payments
    ]
